### PR TITLE
ddl, sessionctx: make auto_random aware of auto_increment_increment

### LINKED ISSUE #16508: UCP: make "AutoRandom" to be `auto_increment_increment`-aware.
## Description

Feature "AutoRandom" is added to TiDB v3.1 and v4.0, as a replacement of "AutoIncrement". This feature aims to provide a better solution to automatically generate primary key ID and avoid hotspots during a heavy write scenario.

For now, as an alternative to generate "auto" keys, "AutoRandom" is not `auto_increment_increment` aware, which lead to the following unexpected results when using JDBC, for example:

```
# Using MySQL client.
tidb> create table t(a bigint key auto_random)
Query OK, 0 rows affected, 1 warning (0.02 sec)
tidb> set global auto_increment_increment=2;                                                           
Query OK, 0 rows affected (0.00 sec)
```
And then,
```
// Scala code
object Test {
  val tidbUrl = s"jdbc:mysql://127.0.0.1:4000/test"

  def main(args: Array[String]): Unit = {
    val conn = DriverManager.getConnection(tidbUrl, username, password)
    val stmt = conn.createStatement()
    val incremental = {
      val res = stmt.executeQuery("select @@auto_increment_increment")
      res.next()
      res.getInt(1)
    }
    println(s"Current auto_increment_increment: $incremental")
    stmt.executeUpdate("insert into t values (), (), (), ()", Statement.RETURN_GENERATED_KEYS)
    val res = stmt.getGeneratedKeys()
    new Iterator[String] {
      def hasNext = res.next()
      def next() = res.getString(1)
    }.toStream.foreach(println)
    stmt.close()
    conn.close()
  }
}
```
Run the codes above, you'll get something like:
```
Current auto_increment_increment: 2
1879048217
1879048219
1879048221
1879048223
```
However, the actual inserted keys are:
```
tidb> select * from t;
+------------+
| a          |
+------------+
| 1879048217 |
| 1879048218 |
| 1879048219 |
| 1879048220 |
+------------+
4 rows in set (0.00 sec)
```

Reason of the wrong answer is that JDBC tries to calculate the generated keys through `LastInsertID()` and `auto_increment_increment`.

You can refer to #14301 to get the way how it is supported by "AutoIncrement".

## Score

* 1680

## Mentor(s)

* [@AilinKid ](https://github.com/AilinKid)
* [@tangenta](https://github.com/tangenta)

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang only.

## Learning Materials
 
* [AUTO_RANDOM column option](https://pingcap.com/docs/dev/reference/sql/attributes/auto-random/)
* [System variable 'auto_increment_increment' and 'auto_increment_offset'](https://dev.mysql.com/doc/refman/8.0/en/replication-options-master.html#sysvar_auto_increment_increment)
* [Closed issue 14245](https://github.com/pingcap/tidb/issues/14245)
* [Retrieving AUTO_INCREMENT Column Values through JDBC](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-usagenotes-last-insert-id.html)

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #16508 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
- validate the value of `auto_increment_increment` and `auto_increment_offset`.
- add a test for `auto_random` value.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note 

- Session variable `auto_increment_increment` and `auto_increment_offset` now affect `auto_random` column value.

### CODE REVIEW

File: ddl/serial_test.go [hunk 1]
```diff
@@ -983,6 +984,66 @@ func (s *testSerialSuite) TestAutoRandom(c *C) {
...
+
+	const truncate, recreate = true, false
+	expect := func(vs ...int) []int { return vs }
+	testCase := []struct {
+		setupAction bool  // truncate or recreate
+		increment   int   // @@auto_increment_increment
+		offset      int   // @@auto_increment_offset
+		results     []int // the implicit allocated auto_random incremental-bit part of values
+	}{
+		{recreate, 5, 10, expect(10, 15, 20)},
```
Comment (djshow832) [inline ddl/serial_test.go:1025]: Any cases to test invalid values? Invalid values should be adjusted to valid values.

Comment (tangenta) [reply ddl/serial_test.go:1025]: Validation check only happens in SetExpr like `set @@auto_increment_increment = 999999`. I add another test. PTAL

**Pull request merged**